RationaleMCP/0027/HindleyMilner.md

```diff
@@ -0,0 +1,201 @@
+
+_If the `e2` is evaluable, and is necessary to evaluate the meta-expression, it should be evaluated._
+
+### Function calls
+* For each expression passed as argument, its unit must be compatible with the declared unit of the corresponding function input.
+* If the unit of the output of the function is `well-formed`, the unit of the expression is the unit of the output of the function.
+* The unit of the expression is `undefined` otherwise.
+
+### Operator der
+The unit of the expression `der(e)` is `der(e.unit)`.
+
+### If expressions
+Consider the expression `if cond then e1 else e2 end if;`
+* If both branches are unit compatible, then the unit of the expression must be compatible with `e1.unit` and with `e2.unit` .
+* If both branches have `well-formed` but incompatible units, the condition must
+  be evaluable and the `if`-expression be replaced by one of the branches. (The
+  unit of the remaining branch will effectively become the unit of the entire
+  `if`-expression.)
+* It is an error otherwise.
+
+### Equations
+Both sides of binding equations, equality equations and connect equations must be unit compatible.
+
+## Unit operators
+
+### `withUnit($value$, $unit$)`
+
+Create an expression with unit `$unit$`, and value `$value$`.
+Argument $value$ needs to be a `Real` expression, with empty unit.
+
+_This operator makes it possible to attach a unit to a literal._
+
+### `withoutUnit($value$, $unit$)`
+Create an expression with empty unit, whose value is the numerical value of `$value$` expressed in `$unit$`.
+Argument `$value$` needs to be a `Real` expression, with a well-formed unit.
+The unit of `$value$` must be compatible with `$unit$`.
+
+### `inUnit($value$, $unit$)`
+Create an expression with unit `$unit$`, whose value is the conversion of `$value$` to `$unit$`.
+Argument `$value$` needs to be a `Real` expression, with a well-formed unit.
+The unit of `$value$` must be compatible with `$unit$`.
+
+## Possible extensions and current limits
+
+### Builtin operators and functions
+The current state of this proposal does not cover the unit checking of the
+different Modelica builtin operators and functions.
+
+### Arrays
+We don't try to specify how this would work an arrays, although the current
+proposal work straight forwardly with element-wise operators. We are confident
+that it could be extended to matrix computation too, but haven't a proof of
+concept for that part.
+
+### Constants
+This proposal take a stance regarding how the unit of constants should be
+handled but we recognise that this is still very much up for discussion. It is
+also somewhat orthogonal to the rest of the proposal.
+
+### Functions
+The Hindler-Milner algorithm, provides a way to infer unit attributes of the
+inputs and outputs of functions. Concretely, it would imply computing the
+bundles of constraints from the body of each functions and adding them to the
+algorithm, after a minimal rewriting, when processing a function call.
```

Comment:
Line breaks:
```suggestion
The Hindler-Milner algorithm, provides a way to infer unit attributes of the inputs and outputs of functions.
Concretely, it would imply computing the bundles of constraints from the body of each functions and adding them to the algorithm, after a minimal rewriting, when processing a function call.
```